File: ts/tests/echo-bot.test.ts
import { jest } from '@jest/globals';
import type { DecryptedMessage, MercuryActivity } from 'webex-message-handler';

const mockSendMessage = jest.fn<(...args: unknown[]) => Promise<Record<string, string>>>()
  .mockResolvedValue({ id: 'msg-1' });
const mockSendCard = jest.fn<(...args: unknown[]) => Promise<Record<string, string>>>()
  .mockResolvedValue({ id: 'msg-1' });

jest.unstable_mockModule('../src/webex', () => ({
  sendMessage: mockSendMessage,
  sendCard: mockSendCard,
  getMessage: jest.fn(),
  getAttachmentAction: jest.fn(),
  extractCards: jest.fn(),
  _setMessagesUrl: jest.fn(),
}));

const { WgrokEchoBot } = await import('../src/echo-bot');
import { loadCases } from './helpers';

interface EchoBotCases {
  config: { domains: string[] };
  routes: Record<string, string>;
  cases: Array<{
    name: string;
    sender: string;
    text: string;
    cards: unknown[];
    expect_send: boolean;
    expected_reply_to?: string;
    expected_reply_text?: string;
    expected_reply_card?: unknown;
    use_routes?: boolean;
  }>;
}

const CASES = loadCases<EchoBotCases>('echo_bot_cases.json');

function fakeMsg(sender: string, text: string): DecryptedMessage {
  return {
    id: 'test-msg-id',
    roomId: 'room-abc',
    personId: 'person-123',
    personEmail: sender,
    text,
    created: new Date().toISOString(),
    raw: {} as MercuryActivity,
  };
}

describe('WgrokEchoBot', () => {
  beforeEach(() => {
    mockSendMessage.mockClear();
    mockSendCard.mockClear();
  });

  it.each(CASES.cases)('$name', async (tc) => {
    const routes = tc.use_routes ? CASES.routes : {};
    const bot = new WgrokEchoBot({
      webexToken: 'fake-token',
      domains: CASES.config.domains,
      debug: false,
      routes,
      platformTokens: { webex: ['fake-token'] },
      webhookPort: null,
      webhookSecret: null,
    });

    const msg = fakeMsg(tc.sender, tc.text);
    await bot.onMessageWithCards(msg, tc.cards);

    if (tc.expect_send) {
      const called = mockSendMessage.mock.calls.length > 0 || mockSendCard.mock.calls.length > 0;
      expect(called).toBe(true);

      if (tc.expected_reply_card) {
        expect(mockSendCard).toHaveBeenCalledTimes(1);
        const [, to, text] = mockSendCard.mock.calls[0] as [string, string, string];
        expect(to).toBe(tc.expected_reply_to);
        expect(text).toBe(tc.expected_reply_text);
      } else {
        expect(mockSendMessage).toHaveBeenCalledTimes(1);
        const [, to, text] = mockSendMessage.mock.calls[0] as [string, string, string];
        expect(to).toBe(tc.expected_reply_to);
        expect(text).toBe(tc.expected_reply_text);
      }
    } else {
      expect(mockSendMessage).not.toHaveBeenCalled();
      expect(mockSendCard).not.toHaveBeenCalled();
    }
  });
});

File: rust/tests/echo_bot_test.rs
use serde::Deserialize;
use serde_json::{json, Value};
use std::collections::HashMap;
use std::fs;
use webex_message_handler::{DecryptedMessage, MercuryActivity};
use wiremock::matchers::{method, path};
use wiremock::{Mock, MockServer, ResponseTemplate};
use wgrok::{BotConfig, WgrokEchoBot, _set_messages_url};

#[derive(Deserialize)]
struct EchoBotCases {
    config: EchoBotConfig,
    routes: HashMap<String, String>,
    cases: Vec<EchoBotCase>,
}

#[derive(Deserialize)]
struct EchoBotConfig {
    domains: Vec<String>,
}

#[derive(Deserialize)]
struct EchoBotCase {
    name: String,
    sender: String,
    text: String,
    cards: Vec<Value>,
    expect_send: bool,
    expected_reply_to: Option<String>,
    expected_reply_text: Option<String>,
    expected_reply_card: Option<Value>,
    #[serde(default)]
    use_routes: bool,
}

fn load_cases() -> EchoBotCases {
    let data = fs::read_to_string("../tests/echo_bot_cases.json").expect("load echo bot cases");
    serde_json::from_str(&data).expect("parse echo bot cases")
}

fn fake_msg(sender: &str, text: &str) -> DecryptedMessage {
    DecryptedMessage {
        id: "test-msg-id".to_string(),
        room_id: "room-abc".to_string(),
        person_id: "person-123".to_string(),
        person_email: sender.to_string(),
        text: text.to_string(),
        html: None,
        created: "2024-01-01T00:00:00Z".to_string(),
        room_type: None,
        raw: MercuryActivity {
            id: String::new(),
            verb: String::new(),
            actor: Default::default(),
            object: Default::default(),
            target: Default::default(),
            published: String::new(),
            encryption_key_url: None,
        },
    }
}

#[tokio::test]
async fn test_echo_bot_cases() {
    let cases = load_cases();

    for tc in &cases.cases {
        let server = MockServer::start().await;
        _set_messages_url(Some(format!("{}/messages", server.uri())));

        Mock::given(method("POST"))
            .and(path("/messages"))
            .respond_with(ResponseTemplate::new(200).set_body_json(json!({"id": "msg-1"})))
            .mount(&server)
            .await;

        let routes = if tc.use_routes {
            cases.routes.clone()
        } else {
            HashMap::new()
        };

        let bot = WgrokEchoBot::new(BotConfig {
            webex_token: "fake-token".to_string(),
            domains: cases.config.domains.clone(),
            debug: false,
            routes,
            platform_tokens: {
                let mut map = HashMap::new();
                map.insert("webex".to_string(), vec!["fake-token".to_string()]);
                map
            },
            webhook_port: None,
            webhook_secret: None,
        });

        let msg = fake_msg(&tc.sender, &tc.text);
        bot.on_message_with_cards(&msg, &tc.cards).await;

        let requests = server.received_requests().await.unwrap();

        if tc.expect_send {
            assert_eq!(
                requests.len(),
                1,
                "case {}: expected 1 request, got {}",
                tc.name,
                requests.len()
            );

            let body: Value = serde_json::from_slice(&requests[0].body).unwrap();
            if let Some(ref expected_to) = tc.expected_reply_to {
                assert_eq!(
                    body["toPersonEmail"], *expected_to,
                    "case {}: reply_to mismatch",
                    tc.name
                );
            }
            if let Some(ref expected_text) = tc.expected_reply_text {
                assert_eq!(
                    body["text"], *expected_text,
                    "case {}: reply_text mismatch",
                    tc.name
                );
            }
            if tc.expected_reply_card.is_some() {
                let has_attachments = body
                    .get("attachments")
                    .and_then(|a| a.as_array())
                    .map_or(false, |a| !a.is_empty());
                assert!(
                    has_attachments,
                    "case {}: expected card attachment",
                    tc.name
                );
            }
        } else {
            assert_eq!(
                requests.len(),
                0,
                "case {}: expected 0 requests, got {}",
                tc.name,
                requests.len()
            );
        }

        _set_messages_url(None);
    }
}
